python/tvm/relay/op/contrib/ethosu.py

```diff
@@ -0,0 +1,251 @@
+            requantize_op = activation.args[0]
+        else:
+            requantize_op = func_body
+        bias_add = requantize_op.args[0]
+        qnn_conv2d = bias_add.args[0]
+        data_layout = qnn_conv2d.attrs.data_layout
+        kernel_layout = qnn_conv2d.attrs.kernel_layout
+        # We consider the weights & biases as params as it should be a Constant
+        self.weights = TensorParams(
+            qnn_conv2d.args[QConv2DArgs.weights.value],
+            kernel_layout,
+            qnn_conv2d.args[QConv2DArgs.weights_scale.value],
+            qnn_conv2d.args[QConv2DArgs.weights_zero_point.value],
+        )
+
+        self.biases = TensorParams(
+            bias_add.args[BiasAddArgs.biases.value],
+            data_layout,
+            requantize_op.args[RequantArgs.ifm_scale.value],
+            requantize_op.args[RequantArgs.ifm_zero_point.value],
+        )
+        self.ifm = TensorParams(
+            qnn_conv2d.args[QConv2DArgs.ifm.value],
+            data_layout,
+            qnn_conv2d.args[QConv2DArgs.ifm_scale.value],
+            qnn_conv2d.args[QConv2DArgs.ifm_zero_point.value],
+        )
+        self.ofm = TensorParams(
+            func_body,
+            data_layout,
+            requantize_op.args[RequantArgs.ofm_scale.value],
+            requantize_op.args[RequantArgs.ofm_zero_point.value],
+        )
+        self.padding = qnn_conv2d.attrs.padding
+        self.strides = qnn_conv2d.attrs.strides
+        self.dilation = qnn_conv2d.attrs.dilation
+        self.activation = activation
+
+        # If groups are equal to channel, its a depthwise_conv2d
+        self.groups = qnn_conv2d.attrs.groups
+        self.is_depthwise = False
+        channels_axis = {"HWIO": 3, "HWOI": 2}
+        if qnn_conv2d.attrs.groups == self.weights.shape[channels_axis[kernel_layout]]:
+            self.is_depthwise = True
+
+    def is_valid(self):
+        """
+        Checks whether QnnConv2D with Clip has compatible attributes with HW
+        """
+        tensor_params = [self.weights, self.ifm, self.ofm]
+        if not check_valid_dtypes(tensor_params):
+            return False
+        if not check_weights(self.weights, self.dilation):
+            return False
+        if not check_bias(self.biases):
+            return False
+        if not check_strides(self.strides):
+            return False
+        if not check_batch_size(self.ifm):
+            return False
+        if not check_dilation(self.dilation):
+            return False
+        if not check_padding(self.padding, self.padding_bounds):
+            return False
+        legal_groups = [1, self.ofm.shape[3]]
+        if self.groups not in legal_groups:
+            return False
+        # This should be a valid QnnDepthwise2DParams, not QnnConv2DParams
+        if self.is_depthwise:
+            return False
+        return True
+
+
+def qnn_conv2d_pattern():
+    """
+    Create pattern for qnn.conv2D with optional fused relu
```

Comment:
Improve the grammar in these if possible.

Reply:
Done